application/data-application.R

```diff
@@ -63,15 +62,103 @@ kokoszka_test <- function(X_fpc, Y_fpc, thre_p = 0.99, thre_q = 0.99) {
                            p = p, q = q, data.name = "Y ~ X"))
   class(result) <- "htest"
   return(result)
+  
+}
 
+# Test for no functional effects by Lee et al. (2020)
+FMDD_test <- function(X_fdata, Y_fdata, int_rule = "trapezoid", B = 1e3) {
+  
+  # Check if X_fdata and Y_fdata are functional data as fdata
+  stopifnot(fda.usc::is.fdata(X_fdata))
+  stopifnot(fda.usc::is.fdata(Y_fdata))
+  
+  # Sample size
+  n <- nrow(X_fdata[["data"]])
+  
+  # Check if sample sizes are matched
+  if (n != nrow(Y_fdata[["data"]])) {
+    stop("X_fdata and Y_fdata samples must have the same number of elements")
+  }
+  
+  # Grid points in which our functional samples are valued
+  s <- X_fdata[["argvals"]]
+  t <- Y_fdata[["argvals"]]
+  
+  # Check if grids are equispaced
+  eps <- sqrt(.Machine[["double.eps"]])
+  equispaced_x <- all(abs(diff(s, differences = 2)) < eps)
+  equispaced_y <- all(abs(diff(t, differences = 2)) < eps)
+  
+  # Building matrices A and B
+  a_ij <- b_ij <- A_mat <- B_mat <- matrix(0, n, n)
+  for (i in 1:n) {
+
+    a_ij[i, ] <- sqrt(apply(t(t(X_fdata[["data"]]) - X_fdata[["data"]][i, ])^2,
+                            1, FUN = "integral1D", s, int_rule, equispaced_x))
+    b_ij[i, ] <- apply(t(t(Y_fdata[["data"]]) - Y_fdata[["data"]][i, ])^2,
+                       1, FUN = "integral1D", t, int_rule, equispaced_y)/2
+  }
+  
+  a_i_dot <- (1 / (n - 2)) * rowSums(a_ij)
+  a_dot_j <- (1 / (n - 2)) * colSums(a_ij)
+  a_dot_dot <- (1 / ((n - 2) * (n - 1))) * sum(sum(a_ij))
+  
+  b_i_dot <- (1 / (n - 2)) * rowSums(b_ij)
+  b_dot_j <- (1 / (n - 2)) * colSums(b_ij)
+  b_dot_dot <- (1 / ((n - 2) * (n - 1))) * sum(sum(b_ij))
+  
+  # A and B matrices: null elements in the diagonal
+  A_mat <- a_ij - matrix(rep(a_i_dot, n), nrow = n) -
```

Comment:
These centering operations can be done more efficiently by implicit column recycling.